File: 1442/solution.py
class Solution:
    def countTriplets(self, arr: list[int]) -> int:
        counts = {0: [0, 1]}
        xor = 0
        res = 0
        for i, v in enumerate(arr):
            xor ^= v
            if xor in counts:
                total_index, count = counts[xor]
                res += (i * count) - total_index
                total_index += i + 1
                count += 1
                counts[xor] = [total_index, count]
            else:
                counts[xor] = [i + 1, 1]
        return res

File: 1442/solution.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public int countTriplets(int[] arr) {
        Map<Integer, int[]> counts = new HashMap<>();
        int xor = 0;
        int res = 0;
        counts.put(0, new int[] { 1, 0 });
        for (int i = 0; i < arr.length; i++) {
            xor ^= arr[i];
            if (counts.containsKey(xor)) {
                int[] value = counts.get(xor);
                int count = value[0];
                int indexTotal = value[1];
                res += (i * count) - indexTotal;
                indexTotal += i + 1;
                count++;
                counts.put(xor, new int[] { count, indexTotal });
            } else {
                counts.put(xor, new int[] { 1, i + 1 });
            }
        }
        return res;
    }
}

// time O(n)
// space O(n)
